selfdrive/locationd/magd.py

```diff
@@ -0,0 +1,221 @@
+    return axes_lengths, rotation_angle
+
+  def get_ellipsoid_center(self, coeffs: np.ndarray) -> np.ndarray:
+    a, b, c, d, e = coeffs[:5]
+    num = (c**2 / 4) - (a * b)
+    x0 = ((b * d / 2) - (c / 2 * e / 2)) / num
+    y0 = ((a * e / 2) - (c / 2 * d / 2)) / num
+    return np.array([x0, y0])
+
+  def estimate_calibration_params(self) -> np.ndarray:
+    points = self.point_buckets.get_points(FIT_POINTS_TOTAL)
+    calibration_params = np.ones(5) * np.nan
+    try:
+      x, y, yaw = points[:, 0], points[:, 1], points[:, 2]
+      D = np.vstack((x * x, y * y, x * y, x, y, np.ones_like(x))).T
+      _, _, V = np.linalg.svd(D, full_matrices=False)
+      coeffs = V[-1] / V[-1, -1]
+      x0, y0 = self.get_ellipsoid_center(coeffs)
+      (l1, l2), theta = self.get_ellipsoid_rotation(coeffs)
+      bearing = self.get_calibrated_bearing(x, y, np.array([x0, y0, l1, l2, theta, 0.0]))
+      offset = np.median(self.reset_angle_range(bearing - yaw))
+      calibration_params = np.array([x0, y0, l1, l2, theta, offset])
+    except np.linalg.LinAlgError as e:
+      cloudlog.exception(f"LinAlgError computing magnetometer calibration params: {e}")
+    except Exception as e:
+      cloudlog.exception(f"Error computing magnetometer calibration params: {e}")
+    return calibration_params
+
+  def reset_angle_range(self, angle: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
+    # reset angle range to [-pi, pi]
+    return (angle + np.pi) % (2 * np.pi) - np.pi
+
+  def get_calibrated_bearing(self, x: Union[float, np.ndarray], y: Union[float, np.ndarray], calibration_params: np.ndarray) -> Any:
+    x0, y0, l1, l2, theta, offset = calibration_params
+    x = x - x0
+    y = y - y0
+    x_new = x * np.cos(theta) + y * np.sin(theta)
+    y_new = -x * np.sin(theta) + y * np.cos(theta)
+    x_new, y_new = x_new / l1, y_new / l2
+    bearing = np.arctan2(y_new, x_new)
+    bearing = self.reset_angle_range(bearing - offset)
+    return bearing
+
+  def handle_log(self, which: str, msg: log.Event) -> None:
+    if which == "carState":
+      self.vego = msg.vEgo
+    elif which == "liveLocationKalman":
+      self.yaw = msg.orientationNED.value[2]
+      self.yaw_valid = msg.orientationNED.valid
+    elif which == "magnetometer":
+      if msg.source == log.SensorEventData.SensorSource.mmc5603nj:
+        raw_vals = np.array(msg.magneticUncalibrated.v)
+        self.raw_vals = (raw_vals[: 3] - raw_vals[3:]) / 2
+        if np.all(np.abs(self.raw_vals - self.past_raw_vals) < 2.0) and msg.magneticUncalibrated.status:
+          self.filtered_vals = np.array([f.update(v) for f, v in zip(self.filtered_magnetic, self.raw_vals, strict=True)])
+        if self.vego > MIN_VEL and self.yaw_valid:
+          self.point_buckets.add_point(self.filtered_vals[0], self.filtered_vals[2], self.yaw)
+        self.past_raw_vals = self.raw_vals
+
+  def get_msg(self, valid: bool = True, with_points: bool = False) -> log.Event:
```

Comment:
`valid=True` is a dangerous default. it's best to pass in every time, otherwise it's easy to write a bug where it's always valid